consensus/src/ledger/merkle.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use std::sync::Arc;
+
+use crate::Ledger;
+use anyhow::*;
+use indexmap::IndexSet;
+use snarkos_storage::Digest;
+use snarkvm_algorithms::MerkleParameters;
+
+use super::indexed_merkle_tree::IndexedMerkleTree;
+
+#[derive(Clone)]
+pub struct MerkleLedger<P: MerkleParameters> {
+    ledger_digests: IndexSet<Digest>,
+    commitments: IndexedMerkleTree<P>,
+    serials: IndexedMerkleTree<P>,
```

Comment:
I've seen `SerialX` and `SerialY` objects added to the codebase, so I'd go with the verbose `serial_numbers` here in order to avoid mixing them together.